### PR TITLE
Factor out dbuf_sync_bonus

### PR DESCRIPTION
**Pull request by mattmacy**
As a small piece of the async COW work refactor part of dbuf_sync_leaf in to
dbuf_sync_bonus and dbuf_undirty_bonus.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Performance enhancement (non-breaking change which improves efficiency)
- [x] Code cleanup (non-breaking change which makes code smaller or more readable)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Documentation (a change to man pages or other documentation)

### Checklist:

- [x] My code follows the ZFS on Linux [code style requirements](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#coding-conventions).
- [x] I have updated the documentation accordingly.
- [ ] I have read the [**contributing** document](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md).
- [ ] I have added [tests](https://github.com/zfsonlinux/zfs/tree/master/tests) to cover my changes.
- [x] All new and existing tests passed.
- [x] All commit messages are properly formatted and contain [`Signed-off-by`](https://github.com/zfsonlinux/zfs/blob/master/.github/CONTRIBUTING.md#signed-off-by).

### CODE REVIEW

**Comment by behlendorf** [comment]
This proposed change introduces the following [bug](http://build.zfsonlinux.org/builders/Ubuntu%2018.04%20x86_64%20%28TEST%29/builds/5215/steps/shell_9/logs/console) which needs to be resolved.

```
[ 8462.658978] BUG: unable to handle kernel NULL pointer dereference at 00000000000000d8
[ 8462.667448] IP: kmem_cache_free+0x11d/0x1e0
[ 8462.671870] PGD 0 P4D 0 
[ 8462.675187] Oops: 0000 [#1] SMP PTI
[ 8462.679065] Modules linked in: zfs(POE) icp(POE) zlua(POE) zcommon(POE) zunicode(POE) znvpair(POE) zavl(POE) spl(OE) nfsd auth_rpcgss nfs_acl lockd grace sunrpc binfmt_misc serio_raw sch_fq_codel ib_iser rdma_cm iw_cm ib_cm ib_core iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi ip_tables x_tables autofs4 btrfs zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear crct10dif_pclmul crc32_pclmul ghash_clmulni_intel pcbc aesni_intel aes_x86_64 crypto_simd glue_helper cryptd [last unloaded: scsi_debug]
[ 8462.713919] CPU: 0 PID: 8661 Comm: dp_sync_taskq Tainted: P           OE    4.15.0-1041-aws #43-Ubuntu
[ 8462.720849] Hardware name: Xen HVM domU, BIOS 4.2.amazon 08/24/2006
[ 8462.725144] RIP: 0010:kmem_cache_free+0x11d/0x1e0
[ 8462.729453] RSP: 0018:ffffb3168b363b58 EFLAGS: 00010286
[ 8462.734620] RAX: 0000000000000000 RBX: ffff98d22be84000 RCX: dead0000000000ff
[ 8462.743156] RDX: 0000000000000000 RSI: ffff98d22be84000 RDI: ffff98d2a2462600
[ 8462.756128] RBP: ffffb3168b363b70 R08: 0000000000000000 R09: 0000000000003f40
[ 8462.768243] R10: 0000000000000000 R11: 00000000000001ee R12: ffff98d2a2462600
[ 8462.777462] R13: ffff98d22be84000 R14: ffff98d2515a8a00 R15: ffff98d26f75c900
[ 8462.783397] FS:  0000000000000000(0000) GS:ffff98d2a7a00000(0000) knlGS:0000000000000000
[ 8462.792404] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[ 8462.799240] CR2: 00000000000000d8 CR3: 0000000073e0a005 CR4: 00000000001606f0
[ 8462.810640] Call Trace:
[ 8462.813405]  spl_kmem_cache_free+0xf1/0x1c0 [spl]
[ 8462.818835]  zio_buf_free+0x58/0x60 [zfs]
[ 8462.823725]  dbuf_undirty_bonus.isra.13+0x41/0x110 [zfs]
[ 8462.828609]  dbuf_sync_bonus.isra.14+0x101/0x2b0 [zfs]
[ 8462.833701]  dbuf_sync_leaf+0x46d/0x780 [zfs]
[ 8462.837761]  ? do_raw_spin_unlock+0x9/0x10 [zfs]
[ 8462.841444]  ? __raw_spin_unlock+0x9/0x10 [zfs]
[ 8462.845458]  ? zfs_space_delta_cb+0xab/0x220 [zfs]
[ 8462.850223]  ? queued_spin_unlock+0x9/0x10 [zfs]
[ 8462.854156]  ? __list_del_entry+0x1d/0x20 [zfs]
[ 8462.857835]  dbuf_sync_list+0xae/0xf0 [zfs]
[ 8462.861269]  dnode_sync+0x667/0xdd0 [zfs]
[ 8462.864623]  ? do_raw_spin_unlock+0x9/0x10 [zfs]
[ 8462.868855]  ? __raw_spin_unlock+0x9/0x10 [zfs]
[ 8462.872537]  dmu_objset_sync_dnodes+0x3b/0x240 [zfs]
[ 8462.876469]  sync_dnodes_task+0x29/0x50 [zfs]
[ 8462.880034]  taskq_thread+0x22d/0x4a0 [spl]
[ 8462.883468]  ? wake_up_q+0x80/0x80
[ 8462.886602]  kthread+0x121/0x140
[ 8462.889618]  ? taskq_thread_spawn+0x50/0x50 [spl]
[ 8462.893312]  ? kthread_create_worker_on_cpu+0x70/0x70
[ 8462.897364]  ret_from_fork+0x35/0x40
```

**Comment by mattmacy** [comment]
> This proposed change introduces the following [bug](http://build.zfsonlinux.org/builders/Ubuntu%2018.04%20x86_64%20%28TEST%29/builds/5215/steps/shell_9/logs/console) which needs to be resolved.
> 
> ```
> [ 8462.658978] BUG: unable to handle kernel NULL pointer dereference at 00000000000000d8
> [ 8462.667448] IP: kmem_cache_free+0x11d/0x1e0
> [ 8462.671870] PGD 0 P4D 0 
> [ 8462.675187] Oops: 0000 [#1] SMP PTI
> [ 8462.679065] Modules linked in: zfs(POE) icp(POE) zlua(POE) zcommon(POE) zunicode(POE) znvpair(POE) zavl(POE) spl(OE) nfsd auth_rpcgss nfs_acl lockd grace sunrpc binfmt_misc serio_raw sch_fq_codel ib_iser rdma_cm iw_cm ib_cm ib_core iscsi_tcp libiscsi_tcp libiscsi scsi_transport_iscsi ip_tables x_tables autofs4 btrfs zstd_compress raid10 raid456 async_raid6_recov async_memcpy async_pq async_xor async_tx xor raid6_pq libcrc32c raid1 raid0 multipath linear crct10dif_pclmul crc32_pclmul ghash_clmulni_intel pcbc aesni_intel aes_x86_64 crypto_simd glue_helper cryptd [last unloaded: scsi_debug]
> [ 8462.713919] CPU: 0 PID: 8661 Comm: dp_sync_taskq Tainted: P           OE    4.15.0-1041-aws #43-Ubuntu
> [ 8462.720849] Hardware name: Xen HVM domU, BIOS 4.2.amazon 08/24/2006
> [ 8462.725144] RIP: 0010:kmem_cache_free+0x11d/0x1e0
> [ 8462.729453] RSP: 0018:ffffb3168b363b58 EFLAGS: 00010286
> [ 8462.734620] RAX: 0000000000000000 RBX: ffff98d22be84000 RCX: dead0000000000ff
> [ 8462.743156] RDX: 0000000000000000 RSI: ffff98d22be84000 RDI: ffff98d2a2462600
> [ 8462.756128] RBP: ffffb3168b363b70 R08: 0000000000000000 R09: 0000000000003f40
> [ 8462.768243] R10: 0000000000000000 R11: 00000000000001ee R12: ffff98d2a2462600
> [ 8462.777462] R13: ffff98d22be84000 R14: ffff98d2515a8a00 R15: ffff98d26f75c900
> [ 8462.783397] FS:  0000000000000000(0000) GS:ffff98d2a7a00000(0000) knlGS:0000000000000000
> [ 8462.792404] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
> [ 8462.799240] CR2: 00000000000000d8 CR3: 0000000073e0a005 CR4: 00000000001606f0
> [ 8462.810640] Call Trace:
> [ 8462.813405]  spl_kmem_cache_free+0xf1/0x1c0 [spl]
> [ 8462.818835]  zio_buf_free+0x58/0x60 [zfs]
> [ 8462.823725]  dbuf_undirty_bonus.isra.13+0x41/0x110 [zfs]
> [ 8462.828609]  dbuf_sync_bonus.isra.14+0x101/0x2b0 [zfs]
> [ 8462.833701]  dbuf_sync_leaf+0x46d/0x780 [zfs]
> [ 8462.837761]  ? do_raw_spin_unlock+0x9/0x10 [zfs]
> [ 8462.841444]  ? __raw_spin_unlock+0x9/0x10 [zfs]
> [ 8462.845458]  ? zfs_space_delta_cb+0xab/0x220 [zfs]
> [ 8462.850223]  ? queued_spin_unlock+0x9/0x10 [zfs]
> [ 8462.854156]  ? __list_del_entry+0x1d/0x20 [zfs]
> [ 8462.857835]  dbuf_sync_list+0xae/0xf0 [zfs]
> [ 8462.861269]  dnode_sync+0x667/0xdd0 [zfs]
> [ 8462.864623]  ? do_raw_spin_unlock+0x9/0x10 [zfs]
> [ 8462.868855]  ? __raw_spin_unlock+0x9/0x10 [zfs]
> [ 8462.872537]  dmu_objset_sync_dnodes+0x3b/0x240 [zfs]
> [ 8462.876469]  sync_dnodes_task+0x29/0x50 [zfs]
> [ 8462.880034]  taskq_thread+0x22d/0x4a0 [spl]
> [ 8462.883468]  ? wake_up_q+0x80/0x80
> [ 8462.886602]  kthread+0x121/0x140
> [ 8462.889618]  ? taskq_thread_spawn+0x50/0x50 [spl]
> [ 8462.893312]  ? kthread_create_worker_on_cpu+0x70/0x70
> [ 8462.897364]  ret_from_fork+0x35/0x40
> ```

should be fixed by replacing zio_buf_free with kmem_free

**Comment by freqlabs** [comment]
I don't think this was supposed to be closed by that commit :)

File: module/zfs/dbuf.c [hunk 1]
```diff
@@ -3975,6 +4029,11 @@ dbuf_sync_leaf_verify_bonus_dnode(dbuf_dirty_record_t *dr)
 	for (; datap_end < datap_max; datap_end++)
 		ASSERT(*datap_end == 0);
 }
+#else
+static void
+dbuf_sync_leaf_verify_bonus_dnode(dbuf_dirty_record_t *dr)
+{
+}
 #endif
```
Comment (ahrens) [inline module/zfs/dbuf.c]: The more idiomatic way of doing this is:
```
static void
dbuf_sync_leaf_verify_bonus_dnode(dbuf_dirty_record_t *dr)
{
#ifdef ZFS_DEBUG
... all the code ...
#endif
}
```

File: module/zfs/dbuf.c [hunk 2]
```diff
@@ -2239,6 +2240,31 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
...
+static void
+dbuf_undirty_bonus(dbuf_dirty_record_t *dr)
+{
+	dmu_buf_impl_t *db = dr->dr_dbuf;
+	int max_bonuslen;
+	struct dnode *dn;
+
+	if (dr->dt.dl.dr_data != db->db.db_data) {
+		dn = DB_DNODE(db);
+		max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
```
Comment (ahrens) [inline module/zfs/dbuf.c]: looks like these can be declared here where they're assigned.

File: module/zfs/dbuf.c [hunk 3]
```diff
@@ -2239,6 +2240,31 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
...
+		kmem_free(dr->dt.dl.dr_data, max_bonuslen);
+		arc_space_return(max_bonuslen, ARC_SPACE_BONUS);
+	}
+	db->db_data_pending = NULL;
+	dr = list_tail(&db->db_dirty_records);
+	list_remove(&db->db_dirty_records, dr);
```
Comment (ahrens) [inline module/zfs/dbuf.c]: Why are we switching from manipulating the passed-in `dr` to the last (earliest) dirty record?  If these are the same, why do the `list_tail()`?

File: module/zfs/dbuf.c [hunk 4]
```diff
@@ -2239,6 +2240,31 @@ dbuf_dirty(dmu_buf_impl_t *db, dmu_tx_t *tx)
...
+	int max_bonuslen;
+	struct dnode *dn;
+
+	if (dr->dt.dl.dr_data != db->db.db_data) {
+		dn = DB_DNODE(db);
+		max_bonuslen = DN_SLOTS_TO_BONUSLEN(dn->dn_num_slots);
+		kmem_free(dr->dt.dl.dr_data, max_bonuslen);
+		arc_space_return(max_bonuslen, ARC_SPACE_BONUS);
+	}
+	db->db_data_pending = NULL;
```
Comment (ahrens) [inline module/zfs/dbuf.c]: Is `db_data_pending` the same as `dr`?  If so, asserting that would be good.

**Comment by behlendorf** [comment]
Closing PR for now. This is an issue we intend to revisit after the initial OpenZFS refactoring is complete.

**Pull request closed without merging**